### DISCUSSION TITLE
Format of .bcs file

### DISCUSSION THREAD

**Discussion by shahul31** [neperfepx/FEPX#88] [Q&A] [answered]
Hello,

Could you please provide me the format of .bcs for applying external boundary conditions, I can see the files for .ori and .phase in examples, but there is no .bcs file for reference.

Thank you.

**Comment by rquey** [comment, marked as answer]
The format is defined in the documentation: https://fepx.info/doc/input.html#externally-defined-boundary-conditions-optional.

**Reply by shahul31** [reply]
Yes, thank you very much @rquey. 
But the explanation is a bit unclear for me to understand "where <node_id> is a unique 1-indexed identification number, <coord_index> defines the sample axis the constraint is applied to, and <vel> is the velocity being applied to the node in the constraint direction. The options for <coord_index> are: x, y, or z. A singular <coord_index>/<vel> pair should be defined per-line for a given <node_id>. The velocities should be prescribed relative to the mesh dimensions and time-step size in order to produce expected strain rates." 
and that is why I have asked for a sample file. 
Are the nodal coordinates are given in <coord_index>?
Thank you in advance.

**Reply by mkasemer** [reply]
At each node you want to place a boundary condition, you tell it the velocities in the x, y, and z sample directions (`<coord_index>`). For example, if node 15 is constrained in the `x` and `y` directions with velocities `1e-2`, while node 273 has boundary conditions placed in the `z` direction with velocity `1e-3`:

```
15 x 1e-2
15 y 1e-2
273 z 1e-3
```

**Reply by shahul31** [reply]
Now it is clear, thank you so much for the explanation. In my case, I am trying to impose boundary conditions on a single grain (one BCC grain surrounded by FCC grains), so I can apply the constraints on the nodes of this single grain is enough right? or do I have to set constraints for every other node also?

**Reply by mkasemer** [reply]
You will also need to constrain some nodes on the boundary of the entire domain (or wherever you wish) to prevent rigid body translation and rotation. It is up to you how this is done. If you intend that the domain's boundary stays a cube, you could constrain all nodes on the boundary in all directions with velocity=0, for example.

**Reply by shahul31** [reply]
Thanks a lot for the clarifications. :)

**Comment by shahul31** [comment]
Hello,
Could you please clarify one more question, I have defined boundary conditions externally with per-node constraints including the BCC grain considered. Is the below line alone should be present in the "## Boundary Condition" section or we can also define other BC types additionally? 
`read_bcs_from_file`

Also, in the .bcs file attached I have made constraints per node for all other nodes in x and y, and for bcc grain nodes I have defined constraints in z (nodes - 6,7,24,33,34,2997,etc) and domain nodes are constrained in all directions with vel=0, but the solver is not converging for elements and solution fails. 

Please clarify this. Many thanks.
[simulation.bcs.txt](https://github.com/neperfepx/FEPX/files/8805121/simulation.bcs.txt)

**Reply by mkasemer** [reply]
@shahul31 When you read BCs from file, by putting the above specified line into your configuration file, it will only consider the BCs as specified in your external file.

Unfortunately, there are many variables that could lead to poor convergence with such a complex applied motion - there is no simple answer to what you are asking. Debugging this will be an arduous task that you have to complete to pinpoint where the issue is arising. I am, again, unsure as to whether you can achieve the overall modeling goal you are attempting using this framework. Further, I am not sure how this deformation modeling will give you insight into the phase transformation problem you are focusing on.

As a note, I do not think having all nodes on the surface of the grain with velocities `x` and `y` equal will achieve what you want. I think you need to have velocities applied to the normal of the surface for the grain, which will likely be very different from node-to-node.

**Reply by shahul31** [reply]
Thank you very much for the clarifications
